### PR TITLE
Feat/admin module sdk47 NTRN-70

### PR DESCRIPTION
**Pull request by quasisamurai**
[TASK](https://hadronlabs.atlassian.net/browse/NTRN-70)

[Admin module PR](https://github.com/neutron-org/admin-module/pull/2)
[Neutron PR](https://github.com/neutron-org/neutron/pull/303)
[Neutron SDK PR](https://github.com/neutron-org/neutron-sdk/pull/107)

### CODE REVIEW

File: x/adminmodule/client/cli/query_archived_proposals.go [hunk 1]
```diff
@@ -40,3 +41,33 @@ func CmdArchivedProposals() *cobra.Command {
 
 	return cmd
 }
+
+func CmdArchivedProposalsLegacy() *cobra.Command {
```
Comment (mmikev) [inline x/adminmodule/client/cli/query_archived_proposals.go:45]: unused method

File: x/adminmodule/types/keys.go [hunk 2]
```diff
@@ -23,12 +23,17 @@ const (
...
 var (
-	ProposalsKeyPrefix        = []byte{0x00}
-	ActiveProposalQueuePrefix = []byte{0x01}
-	ProposalIDKey             = []byte{0x03}
+	ProposalsKeyPrefix              = []byte{0x00}
+	ProposalsKeyLegacyPrefix        = []byte{0x01}
+	ActiveProposalQueuePrefix       = []byte{0x02}
+	ActiveProposalLegacyQueuePrefix = []byte{0x03}
+	ProposalIDKey                   = []byte{0x04}
+	ProposalIDKeyLegacy             = []byte{0x05}
```
Comment (mmikev) [inline x/adminmodule/types/keys.go]: These changes will break the state since you change keys but you don't migrate the storage.
Either implement a migration of the storage, or leave the old keys unchanged

**Review by mmikev — CHANGES_REQUESTED** [review]

File: x/adminmodule/keeper/msg_server_submit_proposal.go [hunk 3]
```diff
@@ -22,22 +20,29 @@ func (k msgServer) SubmitProposal(goCtx context.Context, msg *types.MsgSubmitPro
...
+		return nil, err
 	}
 
-	proposal, err := k.Keeper.SubmitProposal(ctx, content)
+	// TODO: whitelisting. do we need it as soon a we execute only wasm proposals this way?
+	//for _, msg := range msgs {
+	//	if !k.Keeper.IsProposalTypeWhitelisted(msg) {
+	//		return nil, errors.New("proposal content is not whitelisted")
+	//	}
+	//}
```
Comment (mmikev) [inline x/adminmodule/keeper/msg_server_submit_proposal.go]: Discussed in dm, this should be removed

**Review by mmikev — CHANGES_REQUESTED** [review]

File: x/adminmodule/abci.go [hunk 4]
```diff
@@ -68,4 +69,62 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
...
+		var tagValue string
+		var (
+			events sdk.Events
+			msg    sdk.Msg
+		)
+
+		// attempt to execute all messages within the passed proposal
+		// Messages may mutate state thus we use a cached context. If one of
+		// the handlers fails, no state mutation is written and the error
+		// message is logged.
```
Comment (sotnikov-s) [inline x/adminmodule/abci.go]: where exactly is the error message logged?

File: x/adminmodule/abci.go [hunk 5]
```diff
@@ -68,4 +69,62 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
 		)
 		return false
 	})
+
+	keeper.IterateActiveProposalsQueue(ctx, func(proposal v1.Proposal) bool {
+		var tagValue string
```
Comment (sotnikov-s) [inline x/adminmodule/abci.go]: unused var

File: x/adminmodule/abci.go [hunk 6]
```diff
@@ -68,4 +69,62 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
...
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				types.EventTypeAdminProposal,
+				sdk.NewAttribute(govtypes.AttributeKeyProposalID, fmt.Sprintf("%d", proposal.Id)),
+				sdk.NewAttribute(govtypes.AttributeKeyProposalResult, tagValue),
+			),
+		)
+		return false
+	})
 }
```
Comment (sotnikov-s) [inline x/adminmodule/abci.go]: very complicated to read multilevel scopes and shadowed error. please consider such a refactoring: moving the proposal messages handling to a separate func

```suggestion
                cacheCtx, writeCache := ctx.CacheContext()
		events, err := handleProposalMsgs(cacheCtx, keeper, proposal)
		if err != nil {
			proposal.Status = v1.StatusFailed
		} else {
			proposal.Status = v1.StatusPassed
			// write state to the underlying multi-store
			writeCache()

			// propagate the msg events to the current context
			ctx.EventManager().EmitEvents(events)
		}

		keeper.SetProposal(ctx, proposal)
		keeper.RemoveFromActiveProposalQueue(ctx, proposal.Id)

		logger.Info(
			"proposal executed",
			"proposal", proposal.Id,
		)

		ctx.EventManager().EmitEvent(
			sdk.NewEvent(
				types.EventTypeAdminProposal,
				sdk.NewAttribute(govtypes.AttributeKeyProposalID, fmt.Sprintf("%d", proposal.Id)),
				sdk.NewAttribute(govtypes.AttributeKeyProposalResult, tagValue),
			),
		)
		return false
	})
}

func handleProposalMsgs(ctx sdk.Context, keeper keeper.Keeper, proposal v1.Proposal) (sdk.Events, error) {
	var events sdk.Events
	messages, err := proposal.GetMsgs()
	if err != nil {
		return nil, fmt.Errorf("failed to get proposal msgs: %w", err)
	}

	for idx, msg := range messages {
		handler := keeper.Router().Handler(msg)

		var res *sdk.Result
		res, err := handler(ctx, msg)
		if err != nil {
			return nil, fmt.Errorf("failed to handle %d msg in proposal %d: %w", idx, proposal.Id, err)
		}
		events = append(events, res.GetEvents()...)
	}
	return events, nil
}
```

File: x/adminmodule/keeper/keeper.go [hunk 7]
```diff
@@ -18,35 +19,49 @@ type (
 		cdc                       codec.Codec
 		storeKey                  storetypes.StoreKey
 		memKey                    storetypes.StoreKey
-		rtr                       govv1types.Router
-		IsProposalTypeWhitelisted func(govv1types.Content) bool
-		// this line is used by starport scaffolding # ibc/keeper/attribute
+		rtr                       govv1beta1types.Router
+		msgServiceRouter          *baseapp.MsgServiceRouter
+		IsProposalTypeWhitelisted func(govv1beta1types.Content) bool
+		IsModuleWhiteliested      func(typeUrl string) bool
```
Comment (sotnikov-s) [inline x/adminmodule/keeper/keeper.go]: why the funcs are public fields? if you want to use the func outside the struct, you better make it private but create a public interface to it. I know it doesn't really matter here in terms of logic and it's most likely simple misprints, but I just scented a subtle code smell :)

File: x/adminmodule/keeper/msg_server_submit_proposal.go [hunk 8]
```diff
@@ -22,22 +20,29 @@ func (k msgServer) SubmitProposal(goCtx context.Context, msg *types.MsgSubmitPro
 		return nil, fmt.Errorf("proposer %s must be admin to submit proposals to admin-module", msg.Proposer)
 	}
 
-	content := msg.GetContent()
-	if !k.Keeper.IsProposalTypeWhitelisted(content) {
-		return nil, errors.New("proposal content is not whitelisted")
+	msgs, err := msg.GetMsgs()
+	if err != nil {
+		return nil, err
```
Comment (sotnikov-s) [inline x/adminmodule/keeper/msg_server_submit_proposal.go]: please provide some info on the returned error. here and in all the other places where a simple `return err` is used

File: x/adminmodule/keeper/msg_server_submit_proposal.go [hunk 9]
```diff
@@ -22,22 +20,29 @@ func (k msgServer) SubmitProposal(goCtx context.Context, msg *types.MsgSubmitPro
...
+
+	proposal, err := k.Keeper.SubmitProposal(ctx, msgs)
 	if err != nil {
 		return nil, err
 	}
 
 	defer telemetry.IncrCounter(1, types.ModuleName, "proposal")
 
-	submitEvent := sdk.NewEvent(types.EventTypeSubmitAdminProposal, sdk.NewAttribute(govtypes.AttributeKeyProposalType, msg.GetContent().ProposalType()))
+	submitEvent := sdk.NewEvent(types.EventTypeSubmitAdminProposal, sdk.NewAttribute(govtypes.AttributeKeyProposalType, "TODO"))
```
Comment (sotnikov-s) [inline x/adminmodule/keeper/msg_server_submit_proposal.go]: `"TODO"`?

File: x/adminmodule/keeper/msg_server_submit_proposal.go [hunk 10]
```diff
@@ -22,22 +20,28 @@ func (k msgServer) SubmitProposal(goCtx context.Context, msg *types.MsgSubmitPro
...
 
-	content := msg.GetContent()
-	if !k.Keeper.IsProposalTypeWhitelisted(content) {
-		return nil, errors.New("proposal content is not whitelisted")
+	msgs, err := msg.GetMsgs()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, msg := range msgs {
```
Comment (mmikev) [inline x/adminmodule/keeper/msg_server_submit_proposal.go:32]: For each message we also should check that signer of `msg` is admin-module itself.

File: x/adminmodule/keeper/msg_server_submit_proposal.go [hunk 11]
```diff
@@ -22,22 +20,28 @@ func (k msgServer) SubmitProposal(goCtx context.Context, msg *types.MsgSubmitPro
...
-	proposal, err := k.Keeper.SubmitProposal(ctx, content)
+	proposal, err := k.Keeper.SubmitProposal(ctx, msgs)
 	if err != nil {
 		return nil, err
 	}
 
 	defer telemetry.IncrCounter(1, types.ModuleName, "proposal")
 
-	submitEvent := sdk.NewEvent(types.EventTypeSubmitAdminProposal, sdk.NewAttribute(govtypes.AttributeKeyProposalType, msg.GetContent().ProposalType()))
+	submitEvent := sdk.NewEvent(types.EventTypeSubmitAdminProposal, sdk.NewAttribute(govtypes.AttributeKeyProposalType, "TODO"))
```
Comment (mmikev) [inline x/adminmodule/keeper/msg_server_submit_proposal.go]: Huh? "TODO"?

**Review by mmikev — CHANGES_REQUESTED** [review]

File: x/adminmodule/abci.go (hunk 5 above)
Comment (quasisamurai) [reply x/adminmodule/abci.go]: this is not true (mb IDE is lying to you?), due this var is used to emit event at the end

Comment (sotnikov-s) [reply x/adminmodule/abci.go]: sorry, I meant it's never reassigned and the emitted event is an empty string

File: x/adminmodule/client/cli/query_archived_proposals.go [hunk 12]
```diff
@@ -40,3 +41,33 @@ func CmdArchivedProposals() *cobra.Command {
...
+	cmd := &cobra.Command{
+		Use:   "archivedproposalslegacy",
+		Short: "Query archived proposals",
+		Args:  cobra.ExactArgs(0),
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			clientCtx, err := client.GetClientTxContext(cmd)
+			if err != nil {
+				return err
```
Comment (sotnikov-s) [inline x/adminmodule/client/cli/query_archived_proposals.go:54]: add more info to the errors pls

File: x/adminmodule/abci.go (hunk 5 above)
Comment (NeverHappened) [reply x/adminmodule/abci.go]: hm, but it's not set anywhere like above:
`tagValue = govtypes.AttributeValueProposalPassed`

File: x/adminmodule/keeper/msg_server_submit_proposal.go [hunk 13]
```diff
@@ -2,42 +2,57 @@ package keeper
...
-	proposal, err := k.Keeper.SubmitProposal(ctx, content)
+	proposal, err := k.Keeper.SubmitProposal(ctx, msgs)
 	if err != nil {
 		return nil, err
 	}
 
 	defer telemetry.IncrCounter(1, types.ModuleName, "proposal")
 
-	submitEvent := sdk.NewEvent(types.EventTypeSubmitAdminProposal, sdk.NewAttribute(govtypes.AttributeKeyProposalType, msg.GetContent().ProposalType()))
+	submitEvent := sdk.NewEvent(types.EventTypeSubmitAdminProposal, sdk.NewAttribute(govtypes.AttributeKeyProposalType, "sdk message proposal"))
```
Comment (mmikev) [inline x/adminmodule/keeper/msg_server_submit_proposal.go]: I don't very like inlined string here. Let's make it a constant maybe?

File: x/adminmodule/client/cli/query_archived_proposals.go (hunk 12 above)
Comment (quasisamurai) [reply x/adminmodule/client/cli/query_archived_proposals.go:54]: Literally any `x/ `(both neutron & non-neutron) cmd function written this way. is there a reason to do this in a single one?

Comment (sotnikov-s) [reply x/adminmodule/client/cli/query_archived_proposals.go:54]: let's do our best to not increase this chaos. I'd like our code to be well-written. I don't ask you to rewrite all `x/` returned errors this way, just to make the ones you write like that

**Comment by NeverHappened** [comment]
is it feasible to fix unit tests for this module? right now i'm getting compilation errors on `make test`

File: x/adminmodule/keeper/proposal.go [hunk 14]
```diff
@@ -7,22 +7,11 @@ import (
...
 )
 
 // SubmitProposal create new proposal given a content
-func (k Keeper) SubmitProposal(ctx sdk.Context, content govv1types.Content) (govv1types.Proposal, error) {
-	if !k.rtr.HasRoute(content.ProposalRoute()) {
-		return govv1types.Proposal{}, sdkerrors.Wrap(govtypes.ErrNoProposalHandlerExists, content.ProposalRoute())
-	}
-
-	cacheCtx, _ := ctx.CacheContext()
-	handler := k.rtr.GetRoute(content.ProposalRoute())
```
Comment (NeverHappened) [inline x/adminmodule/keeper/proposal.go:22]: why did you remove this cached call? This allowed us to get the error message on unsuccessful handler call

Comment (NeverHappened) [reply x/adminmodule/keeper/proposal.go:22]: ah, have the solution here https://github.com/cosmos/cosmos-sdk/pull/10868#discussion_r784872842

File: x/adminmodule/client/cli/tx_submit_proposal.go [hunk 15]
```diff
@@ -79,7 +79,7 @@ $ %s tx adminmodule submit-proposal --title="Test Proposal" --description="My aw
 
 			content, _ := govv1types.ContentFromProposalType(proposal.Title, proposal.Description, proposal.Type)
 
-			msg, err := types.NewMsgSubmitProposal(content, clientCtx.GetFromAddress())
+			msg, err := types.NewMsgSubmitProposalLegacy(content, clientCtx.GetFromAddress())
```
Comment (sotnikov-s) [inline x/adminmodule/client/cli/tx_submit_proposal.go:82]: not legacy proposal submission won't be available via cli anymore. why?

Comment (quasisamurai) [reply x/adminmodule/client/cli/tx_submit_proposal.go:82]: because it is basically useless: the only way neutron accepts proposal submission (both legacy & non legacy) is via DAO

File: x/adminmodule/types/keys.go [hunk 16]
```diff
@@ -23,12 +23,17 @@ const (
 	AdminKey = "Admin-"
 
 	ArchiveKey = "Archive-"
+
+	ArchiveLegacyKey = "ArchiveLegacy-"
```
Comment (swelf19) [inline x/adminmodule/types/keys.go]: seems like state breaking change. I mean, do we need to move all ArchivedProposals from `Archive-` to `ArchiveLegacy-` and clear `Archive-`

File: x/adminmodule/abci.go [hunk 17]
```diff
@@ -47,13 +48,13 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
 			logMsg = fmt.Sprintf("proposal failed on execution: %s", err)
 		}
 
-		keeper.SetProposal(ctx, proposal)
-		keeper.RemoveFromActiveProposalQueue(ctx, proposal.ProposalId)
+		keeper.SetProposalLegacy(ctx, proposal)
+		keeper.RemoveFromActiveProposalQueueLegacy(ctx, proposal.ProposalId)
 
-		keeper.AddToArchive(ctx, proposal)
+		keeper.AddToArchiveLegacy(ctx, proposal)
```
Comment (swelf19) [inline x/adminmodule/abci.go]: we archive legacy, but do not archive `sdk47-style`, is it intended to be.

File: x/adminmodule/keeper/msg_server_submit_proposal.go [hunk 18]
```diff
@@ -2,42 +2,57 @@ package keeper
...
-		return nil, errors.New("proposal content is not whitelisted")
+	msgs, err := msg.GetMsgs()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to submit proposal")
+	}
+
+	for _, msg := range msgs {
+		signers := msg.GetSigners()
+		if len(signers) != 1 {
+			return nil, fmt.Errorf("should be only 1 signer in message, recieved: %s", msg.GetSigners())
```
Comment (NeverHappened) [inline x/adminmodule/keeper/msg_server_submit_proposal.go]: typo: recieved -> received

File: x/adminmodule/keeper/proposal.go (hunk 14 above)
Comment (NeverHappened) [reply x/adminmodule/keeper/proposal.go:22]: @pr0n00gler are we gonna implement the call of handler right away instead of waiting for the Endblocker?

File: x/adminmodule/types/keys.go (hunk 16 above)
Comment (quasisamurai) [reply x/adminmodule/types/keys.go]: No. As soon as legacy proposals has the same key & and structure as before, we change only keys in terms of golang variables

File: x/adminmodule/client/cli/tx_submit_proposal.go (hunk 15 above)
Comment (sotnikov-s) [reply x/adminmodule/client/cli/tx_submit_proposal.go:82]: do we need the submit legacy proposal at all? aren't we safe to remove it?

File: x/adminmodule/keeper/proposal_legacy.go [hunk 19]
```diff
@@ -29,16 +28,36 @@ func (k Keeper) SubmitProposalLegacy(ctx sdk.Context, content govv1beta1types.Co
...
+	// is written and the error message is returned.
+	err = handler(cacheCtx, content)
+	if err == nil {
+		proposal.Status = govv1beta1types.StatusPassed
+
+		// The cached context is created with a new EventManager. However, since
+		// the proposal handler execution was successful, we want to track/keep
+		// any events emitted, so we re-emit to "merge" the events into the
+		// original Context's EventManager.
+		ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
```
Comment (sotnikov-s) [inline x/adminmodule/keeper/proposal_legacy.go]: this seems not to be necessary because this is handled in the writeCache func:
image

File: x/adminmodule/abci.go [hunk 20]
```diff
@@ -95,6 +95,7 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
 
 		keeper.SetProposal(ctx, proposal)
 		keeper.RemoveFromActiveProposalQueue(ctx, proposal.Id)
+		keeper.AddToArchive(ctx, proposal)
```
Comment (sotnikov-s) [inline x/adminmodule/abci.go]: why do we still need the active proposals handling in the EndBlock? I thought, since we switch to instant proposals execution, we can remove this logic from the adminmodule's EndBlock

File: x/adminmodule/keeper/proposal_legacy.go [hunk 21]
```diff
@@ -29,16 +28,36 @@ func (k Keeper) SubmitProposalLegacy(ctx sdk.Context, content govv1beta1types.Co
...
+
+		// write state to the underlying multi-store
+		writeCache()
+	} else {
+		return govv1beta1types.Proposal{}, err
```
Comment (sotnikov-s) [inline x/adminmodule/keeper/proposal_legacy.go]: so did I get it right that we won't have any proposals in the adminmodule state but the `Passed` ones? the failed proposals won't be recorded because such transactions won't succeed and no state mutations will happen at the adminmodule side, all of them will only be visible from the Neutron DAO's side?

File: x/adminmodule/abci.go [hunk 22]
```diff
@@ -18,10 +19,10 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
 
 	logger := keeper.Logger(ctx)
 
-	keeper.IterateActiveProposalsQueue(ctx, func(proposal govv1beta1types.Proposal) bool {
+	keeper.IterateActiveProposalsQueueLegacy(ctx, func(proposal govv1beta1types.Proposal) bool {
```
Comment (swelf19) [inline x/adminmodule/abci.go]: Why didnt we remove an iteration at all, since we move proposal execution to `SubmitProposal`, the queue is always empty.

File: x/adminmodule/abci.go [hunk 23]
```diff
@@ -1,71 +1,15 @@
...
 	"github.com/cosmos/cosmos-sdk/telemetry"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
-	govv1beta1types "github.com/cosmos/cosmos-sdk/x/gov/types/v1beta1"
 )
 
-// EndBlocker called every block, process inflation, update validator set.
-func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
+// EndBlocker called every block
+func EndBlocker(_ sdk.Context, _ keeper.Keeper) {
```
Comment (sotnikov-s) [inline x/adminmodule/abci.go]: I think we're safe to remove this func now when we removed all the EndBlock proposals processing logic

File: x/adminmodule/types/keys.go [hunk 24]
```diff
@@ -48,7 +53,17 @@ func ProposalKey(proposalID uint64) []byte {
...
+	return append(ProposalsKeyLegacyPrefix, GetProposalIDBytes(proposalID)...)
+}
+
 // ActiveProposalQueueKey returns the key for a proposalID in the activeProposalQueue
 func ActiveProposalQueueKey(proposalID uint64) []byte {
 	return append(ActiveProposalQueuePrefix, GetProposalIDBytes(proposalID)...)
 }
+
+// ActiveProposalLegacyQueueKey returns the key for a proposalID in the activeProposalLegacyQueue
+func ActiveProposalLegacyQueueKey(proposalID uint64) []byte {
```
Comment (sotnikov-s) [inline x/adminmodule/types/keys.go]: both active proposals related funcs (as well as respective keys) are not needed anymore

File: x/adminmodule/types/message_submit_proposal.go [hunk 25]
```diff
@@ -20,37 +15,31 @@ var (
...
 
-func (m *MsgSubmitProposal) GetContent() govtypesv1b1.Content {
-	content, ok := m.Content.GetCachedValue().(govtypesv1b1.Content)
-	if !ok {
-		return nil
-	}
-	return content
+// GetMsgs unpacks m.Messages Any's into sdk.Msg's
+func (m *MsgSubmitProposal) GetMsgs() ([]sdk.Msg, error) {
+	return sdktx.GetMsgs(m.Messages, "sdk.MsgProposal")
```
Comment (sotnikov-s) [inline x/adminmodule/types/message_submit_proposal.go]: sdk.MsgProposal  -> sdk.Msg?

File: x/adminmodule/types/message_submit_proposal.go [hunk 26]
```diff
@@ -77,22 +67,9 @@ func (m *MsgSubmitProposal) ValidateBasic() error {
 	if m.Proposer == "" {
 		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, m.Proposer)
 	}
-
```
Comment (sotnikov-s) [inline x/adminmodule/types/message_submit_proposal.go:80]: wyt about adding a simple validation here e.g. len(m.Messages) > 0, m.Messages are of sdk.Msg type, maybe something else?

File: x/adminmodule/types/message_submit_proposal_legacy.go [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+	if !ok {
+		return nil
+	}
+	return content
+}
+
+func (m *MsgSubmitProposalLegacy) SetContent(content govtypesv1b1.Content) error {
+	msg, ok := content.(proto.Message)
+	if !ok {
+		return fmt.Errorf("can't proto marshal %T", msg)
```
Comment (sotnikov-s) [inline x/adminmodule/types/message_submit_proposal_legacy.go]: can't proto marshal? I think it would be better to phrase it like "failed to cast proposal content of type %T to proto.Message" or something

File: x/adminmodule/types/message_submit_proposal_legacy.go [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+	return sdk.MustSortJSON(bz)
+}
+
+// String implements the Stringer interface
+func (m *MsgSubmitProposalLegacy) String() string {
+	out, _ := yaml.Marshal(m)
+	return string(out)
+}
+
+// ValidateBasic implements Msg
```
Comment (sotnikov-s) [inline x/adminmodule/types/message_submit_proposal_legacy.go]: ValidateBasic **implements Msg**?

File: x/adminmodule/types/message_submit_proposal_legacy.go [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+func (m *MsgSubmitProposalLegacy) ValidateBasic() error {
+	if m.Proposer == "" {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, m.Proposer)
```
Comment (sotnikov-s) [inline x/adminmodule/types/message_submit_proposal_legacy.go]: what's the point of using an empty address as a description? I'd rather put there something like "empty proposer field"

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

File: x/adminmodule/client/cli/tx_submit_proposal.go (hunk 15 above)
Comment (quasisamurai) [reply x/adminmodule/client/cli/tx_submit_proposal.go:82]: we are not safe to remove it due some modules hasn't updated for sdk-47

File: x/adminmodule/types/message_submit_proposal_legacy.go (hunk 29 above)
Comment (quasisamurai) [reply x/adminmodule/types/message_submit_proposal_legacy.go]: This is the way how it's done in current admin module & `x/gov` in sdk. So there his no point to rewrite _legacy_ code

File: x/adminmodule/types/message_submit_proposal_legacy.go (hunk 27 above)
Comment (quasisamurai) [reply x/adminmodule/types/message_submit_proposal_legacy.go]: This is the way how it's done in current admin module & `x/gov `in sdk. So there his no point to rewrite _legacy_ code

File: x/adminmodule/keeper/keeper.go [hunk 30]
```diff
@@ -18,35 +19,59 @@ type (
...
 	cdc codec.Codec,
 	storeKey,
 	memKey storetypes.StoreKey,
-	rtr govv1types.Router,
-	isProposalTypeWhitelisted func(govv1types.Content) bool,
-	// this line is used by starport scaffolding # ibc/keeper/parameter
+	rtr govv1beta1types.Router,
+	msgServiceRouter *baseapp.MsgServiceRouter,
+	isProposalTypeWhitelisted func(govv1beta1types.Content) bool,
+	IsMessageWhitelisted func(msg sdk.Msg) bool,
```
Comment (NeverHappened) [inline x/adminmodule/keeper/keeper.go]: should be `isMessageWhitelisted`

File: x/adminmodule/keeper/keeper.go [hunk 31]
```diff
@@ -18,35 +19,59 @@ type (
 		cdc                       codec.Codec
 		storeKey                  storetypes.StoreKey
 		memKey                    storetypes.StoreKey
-		rtr                       govv1types.Router
-		IsProposalTypeWhitelisted func(govv1types.Content) bool
-		// this line is used by starport scaffolding # ibc/keeper/attribute
+		rtr                       govv1beta1types.Router
```
Comment (NeverHappened) [inline x/adminmodule/keeper/keeper.go]: maybe rename rtr -> legacyRouter for clarity?

File: x/adminmodule/keeper/proposal.go [hunk 32]
```diff
@@ -7,39 +7,38 @@ import (
...
-	if err := handler(cacheCtx, content); err != nil {
-		return govv1types.Proposal{}, sdkerrors.Wrap(govtypes.ErrInvalidProposalContent, err.Error())
-	}
-
+func (k Keeper) SubmitProposal(ctx sdk.Context, msgs []sdk.Msg) (govv1types.Proposal, error) {
```
Comment (NeverHappened) [inline x/adminmodule/keeper/proposal.go:13]: I'm proposing to create another task not to burned Pasha, but we need to rename SubmitProposal() -> ExecuteProposal() and SubmitLegacyProposal() -> ExecuteLegacyProposal()
Because now it's very misleading to read

File: x/adminmodule/keeper/proposal_archive_legacy.go [hunk 33]
```diff
@@ -0,0 +1,34 @@
+package keeper
+
+import (
+	"github.com/cosmos/admin-module/x/adminmodule/types"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	govv1beta1types "github.com/cosmos/cosmos-sdk/x/gov/types/v1beta1"
+)
+
+func (k Keeper) GetArchivedProposalsLegacy(ctx sdk.Context) []*govv1beta1types.Proposal {
```
Comment (NeverHappened) [inline x/adminmodule/keeper/proposal_archive_legacy.go:10]: Not to burden Pasha again - how about create new task to create unit tests for Keeper methods.
It's really easy to make a mistake here with all keys and stuff

**Review by NeverHappened — APPROVED** [review]
LGTM! Few minor comments.
I also would appreciate create separate tasks for keeper unit tests and renaming SubmitProposal names everywhere

File: x/adminmodule/types/message_submit_proposal_legacy.go (hunk 29 above)
Comment (sotnikov-s) [reply x/adminmodule/types/message_submit_proposal_legacy.go]: I strongly disagree with this statement and I find it distructive. It doesn't matter what is the origin of a bad piece of code, I think we are responsible for making it good as long as we own and maintain it.

Comment (quasisamurai) [reply x/adminmodule/types/message_submit_proposal_legacy.go]: anyway I fixed that c:

**Review by mmikev — APPROVED** [review]

**Review by sotnikov-s — APPROVED** [review]

File: x/adminmodule/keeper/keeper.go [hunk 34]
```diff
@@ -18,35 +19,59 @@ type (
...
 )
 
 func NewKeeper(
 	cdc codec.Codec,
 	storeKey,
 	memKey storetypes.StoreKey,
-	rtr govv1types.Router,
-	isProposalTypeWhitelisted func(govv1types.Content) bool,
-	// this line is used by starport scaffolding # ibc/keeper/parameter
+	rtr govv1beta1types.Router,
```
Comment (NeverHappened) [inline x/adminmodule/keeper/keeper.go]: legacyRouter ?

File: x/adminmodule/keeper/proposal_legacy.go [hunk 35]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+// SetProposalLegacy set a proposal to store
+func (k Keeper) SetProposalLegacy(ctx sdk.Context, proposal govv1beta1types.Proposal) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.MustMarshalProposalLegacy(proposal)
+
+	store.Set(types.ProposalLegacyKey(proposal.ProposalId), bz)
+
```
Comment (NeverHappened) [inline x/adminmodule/keeper/proposal_legacy.go]: space left
